dali/kernels/signal/moving_mean_square_gpu.cu

```diff
@@ -0,0 +1,376 @@
+  using type = double;
+};
+
+template <typename T>
+using acc_t = typename accumulator_type<T>::type;
+
+}  // namespace
+
+
+struct SampleDesc {
+  void *out;
+  const void *in;
+  int64_t len;
+};
+
+struct conflict_free_pos {
+  DALI_HOST_DEV DALI_FORCEINLINE
+  int operator()(int pos) const noexcept {
+    return pos + CONFLICT_FREE_OFFSET(pos);
+  }
+};
+
+struct square {
+  template <typename T>
+  DALI_HOST_DEV DALI_FORCEINLINE
+  T operator()(T x) const noexcept {
+    return x * x;
+  }
+};
+
+template <typename T>
+struct divide {
+  explicit divide(T divisor) : divisor_(divisor) {}
+
+  DALI_HOST_DEV DALI_FORCEINLINE
+  T operator()(T x) const noexcept {
+    return x / divisor_;
+  }
+
+  T divisor_;
+};
+
+
+template <typename T, typename SharedMemPos = conflict_free_pos>
+__device__ void PrefixSumSharedMem(T* buffer, int pow2,
+                                   SharedMemPos shm_pos = {}) {
+  int offset = 1;
+  int tid = threadIdx.x;
+  for (int d = pow2 >> 1; d > 0; d >>= 1) {  // build sum in place up the tree
+    __syncthreads();
+    if (tid < d) {
+      int ai = offset * (2 * tid + 1) - 1;
+      int bi = offset * (2 * tid + 2) - 1;
+      buffer[shm_pos(bi)] += buffer[shm_pos(ai)];
+    }
+    offset <<= 1;
+  }
+
+  if (tid == 0) {
+    int last = pow2 - 1;
+    buffer[shm_pos(last)] = 0;  // clear the last element
+  }
+
+  for (int d = 1; d < pow2; d <<= 1) {  // traverse down tree & build scan
+    offset >>= 1;
+    __syncthreads();
+    if (tid < d) {
+      int shm_pos_ai = shm_pos(offset * (2 * tid + 1) - 1);
+      int shm_pos_bi = shm_pos(offset * (2 * tid + 2) - 1);
+      auto t = buffer[shm_pos_ai];
+      buffer[shm_pos_ai] = buffer[shm_pos_bi];
+      buffer[shm_pos_bi] += t;
+    }
+  }
+}
+
+template <typename T, bool inclusive = true, typename Preprocessor = dali::identity,
+          typename Postprocessor = dali::identity, typename SharedMemPos = conflict_free_pos>
+__global__ void BlockPrefixSum(const SampleDesc *samples, int pow2,
+                               Preprocessor pre = {}, Postprocessor post = {},
+                               SharedMemPos shm_pos = {}) {
+  extern __shared__ acc_t<T> temp[];  // allocated on invocation
+
+  int sample_idx = blockIdx.y;
+
+  auto &sample = samples[sample_idx];
+  T *output = reinterpret_cast<T *>(sample.out);
+  const T *input = reinterpret_cast<const T *>(sample.in);
+  int64_t sample_len = sample.len;
+  int64_t grid_stride = gridDim.x * blockDim.x;
+
+  for (int64_t sample_offset = static_cast<int64_t>(blockIdx.x) * blockDim.x;
+       sample_offset < sample_len; sample_offset += grid_stride) {
+    __syncthreads();
+
+    const T *in_ptr = input + sample_offset;
+    T *out_ptr = output + sample_offset;
+
+    int n = cuda_min(static_cast<int>(sample.len - sample_offset), static_cast<int>(blockDim.x));
+
+    for (int pos = threadIdx.x; pos < n; pos += blockDim.x) {
+      temp[shm_pos(pos)] = pre(in_ptr[pos]);
+    }
```

Comment:
Don't we want to zero-pad the temp buffer?

Reply:
Absolutely. Fixed.